ai_flow/task_executor/kubernetes/kube_scheduler.py

```diff
@@ -0,0 +1,126 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import json
+import logging
+from multiprocessing import Process
+from queue import Queue
+
+from ai_flow.common.exception.exceptions import AIFlowConfigException
+from kubernetes.client import V1PodList, V1Pod
+
+from ai_flow.model.task_execution import TaskExecutionKey
+from kubernetes import client
+from kubernetes.client.rest import ApiException
+
+from ai_flow.model.status import TaskStatus
+from ai_flow.task_executor.kubernetes.helpers import create_pod_id, key_to_label_selector, POISON, \
+    gen_command
+from ai_flow.task_executor.kubernetes.kube_config import KubeConfig
+from ai_flow.task_executor.kubernetes.pod_generator import PodGenerator
+
+logger = logging.getLogger(__name__)
+
+
+class KubernetesScheduler(Process):
+
+    def __init__(
+        self,
+        kube_config: KubeConfig,
+        task_queue: 'Queue[TaskExecutionKey]',
+        kube_client: client.CoreV1Api,
+    ):
+        self.kube_config = kube_config
+        self.task_queue = task_queue
+        self.kube_client = kube_client
+        self.namespace = self.kube_config.get_namespace()
+        super().__init__(target=self.schedule)
+
+    def schedule(self):
+        while True:
+            try:
+                key = self.task_queue.get()
+                if key is POISON:
+                    logger.info("Receiving the poison, breaking the loop...")
+                    break
+
+                label_selector = key_to_label_selector(key)
+                if len(self.list_pods(label_selector)) > 0:
+                    logger.warning(f'TaskExecution: {key} has been submitted in the past, skipping...')
+                    continue
+
+                logger.debug(f'Submitting TaskExecution: {key}.')
+                self._run(key)
+            except ApiException as e:
+                if e.reason == "BadRequest":
+                    logger.error("Request was invalid. Failing task")
+                    # TODO call scheduler interface to save meta
+                    print("Save task execution {} with status {} to meta".format(key, TaskStatus.FAILED))
+                else:
+                    logger.warning(
+                        'ApiException when attempting to run task, re-queueing. Message: %s',
+                        json.loads(e.body)['message'],
+                    )
+                    self.task_queue.put(key)
```

Comment:
Insert into the head position of the task_queue.

Reply:
This queue is for starting tasks only, so i think the failed key can be put to the tail of the queue